docs/proposals/auth.md

```diff
@@ -0,0 +1,150 @@
+separating authorization for model and Oldtimer queries from authorization to
+push metrics.  However, this is suboptimal, since it also separates
+authentication.
+
+Requirements
+------------
+
+Heapster should be able to separate authentication from authorization, and
+the authorization interface should be flexible enough to support authorization
+implementations that are able to make the following distinctions:
+
+- containing queries to a particular namespace: we should be able to only allow
+  pods to query for metrics within their own namespace.
+
+- separating querying from pushing: we should allow certain users to push, and
+  certain users to query, but not necessarily tie the two together
+
+- allowing different push scope permissions: we should be able to specify that
+  some users are only allowed to push metrics for the pods within their own
+  RC or namespace, while others should be allowed to push for any object in
+  the cluster.
+
+Proposed Solution
+-----------------
+
+Like in Kubernetes, authentication and authorization should be split into
+separate interfaces.  Each would be configured using a URL flag at the command
+line, similarly to how sinks and sources are currently configured.  Adding
+multiple authentication drivers would be supported (for example, to support
+authenticating from both client certificates and bearer tokens).
+
+### Authentication ###
+
+For authentication, the Kubernetes interface
+(`"k8s.io/kubernetes/pkg/auth/authenticator".Request`) can be reused so that
+any authentication plugins for Kubernetes can be easily used with Heapster.
+The interface is generic enough that it could also be used for specialized
+Heapster-specific interfaces.
+
+For example, if it was desired to use client certificate authentication,
+the command line flag might look like
+`--authn=clientca:/etc/heapster/ca.crt`.
+
+### Authorization ###
+
+For authorization, a system tailored to Heapster directly would be more useful:
+
+```go
+// This is basically just core.HistoricalKey from the Oldtimer PR,
+// renamed to be more generic
+type MetricKey struct {
+    // One of the core.MetricSetType* labels
+    ObjectType string
+
+    NodeName string
+    NamespaceName string
+    PodName string
+    ContainerName string
+    PodId string
+}
+
+type HeapsterQuerySource string
+var (
+    QuerySourceModel HeapsterQuerySource = "model"
+    QuerySourceHistorical HeapsterQuerySource = "historical"
+)
+
+type Authorizer interface {
+    // AuthorizeCreation authorizes creation of metrics with the given prefix
+    // (if any) and the given names at the given keys.
+    AuthorizeCreation(user kauthuser.Info, keys []MetricKey, prefix string, metricNames []string) error 
+
+    // AuthorizeQuery authorizes querying of metrics at the given keys with the
+    // given names from the given source.
+    AuthorizeQuery(user kauthuser.Info, source HeapsterQuerySource, keys []MetricKey, metricNames []string) error
```

Comment:
Partially -- create is POST, but the URL doesn't say anything about the particular metrics, resources, etc involved (and multiple metrics and resources may be involved).

Query is GET, and the URL does say something about the resources involved.

However, that seems like a detail for the particular auth mechanism.  An authorization mechanism like the current one (just using a list of user names) probably doesn't care about GET vs POST, and just wants to know which action (create vs query model vs query historical) is being performed, so it can check the list of users allowed.